File: services/eval/eval.go
// Package eval provides evaluation and quality testing for LLM outputs.
package eval

import (
	"time"
)

// EvalRunStatus represents the status of an evaluation run.
type EvalRunStatus int

const (
	EvalRunStatusUnspecified EvalRunStatus = iota
	EvalRunStatusPending
	EvalRunStatusRunning
	EvalRunStatusCompleted
	EvalRunStatusFailed
	EvalRunStatusCancelled
)

// EvalRun represents an evaluation execution.
type EvalRun struct {
	ID                string
	Name              string
	Description       string
	PromptID          string
	PromptVersion     int
	DatasetID         string
	Config            EvalConfig
	Status            EvalRunStatus
	ErrorMessage      string
	TotalExamples     int
	CompletedExamples int
	Summary           *EvalSummary
	CreatedAt         time.Time
	StartedAt         *time.Time
	CompletedAt       *time.Time
	CreatedBy         string
	Metadata          map[string]string
}

// EvalConfig contains configuration for an evaluation run.
type EvalConfig struct {
	Evaluators  []EvaluatorConfig
	Provider    string
	Model       string
	Concurrency int
	SampleSize  int
	Shuffle     bool
}

// EvaluatorConfig contains configuration for a single evaluator.
type EvaluatorConfig struct {
	Type   string // exact_match, semantic_similarity, llm_judge, custom
	Name   string
	Params map[string]string
	Weight float64
}

// EvalSummary contains aggregated results from an evaluation.
type EvalSummary struct {
	OverallScore      float64
	ScoresByEvaluator map[string]float64
	PassedCount       int
	FailedCount       int
	PassRate          float64
	TotalCostUSD      float64
	TotalTokens       int
	AvgLatencyMs      float64
}

// EvalResult represents a single example's evaluation result.
type EvalResult struct {
	ID               string
	EvalRunID        string
	ExampleID        string
	Input            map[string]interface{}
	ExpectedOutput   map[string]interface{}
	ActualOutput     map[string]interface{}
	EvaluatorResults map[string]*EvaluatorResult
	OverallScore     float64
	Passed           bool
	LatencyMs        float64
	TokensUsed       int
	CostUSD          float64
	Error            string
}

// EvaluatorResult contains the result from a single evaluator.
type EvaluatorResult struct {
	EvaluatorType string
	Score         float64
	Passed        bool
	Explanation   string
	Details       map[string]string
}

// Evaluator describes an available evaluator type.
type Evaluator struct {
	Type        string
	Name        string
	Description string
	Params      []EvaluatorParam
}

// EvaluatorParam describes a parameter for an evaluator.
type EvaluatorParam struct {
	Name         string
	Type         string
	Description  string
	Required     bool
	DefaultValue string
}

// CreateEvalRunInput contains input for creating an evaluation run.
type CreateEvalRunInput struct {
	Name          string
	Description   string
	PromptID      string
	PromptVersion int
	DatasetID     string
	Config        EvalConfig
	Metadata      map[string]string
	CreatedBy     string
}

// ListEvalRunsQuery contains filters for listing evaluation runs.
type ListEvalRunsQuery struct {
	PromptID  string
	DatasetID string
	Status    EvalRunStatus
	Limit     int
	Offset    int
}

// GetEvalResultsQuery contains filters for getting evaluation results.
type GetEvalResultsQuery struct {
	EvalRunID  string
	FailedOnly bool
	Limit      int
	Offset     int
}

// RunComparison contains summary data for comparing runs.
type RunComparison struct {
	RunID         string
	PromptVersion string
	OverallScore  float64
	PassRate      float64
	AvgLatencyMs  float64
	TotalCostUSD  float64
}

// ExampleComparison compares a single example across two runs.
type ExampleComparison struct {
	ExampleID  string
	ScoreA     float64
	ScoreB     float64
	ScoreDiff  float64
	Regression bool
}

// CompareRunsResult contains the full comparison between two runs.
type CompareRunsResult struct {
	RunA         RunComparison
	RunB         RunComparison
	ScoreDiff    float64
	Regressions  int
	Improvements int
	Examples     []ExampleComparison
}

File: sdk/python/src/delos/models/eval.py
"""Models for the eval service."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel


class EvalRunStatus(str, Enum):
    """Status of an evaluation run."""

    UNSPECIFIED = "unspecified"
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class EvaluatorConfig(BaseModel):
    """Configuration for a single evaluator."""

    type: str  # exact_match, semantic_similarity, llm_judge, etc.
    name: str = ""
    params: dict[str, str] = {}
    weight: float = 1.0


class EvalConfig(BaseModel):
    """Configuration for an evaluation run."""

    evaluators: list[EvaluatorConfig] = []
    provider: str = ""
    model: str = ""
    concurrency: int = 1
    sample_size: int = 0  # 0 = all examples
    shuffle: bool = False


class EvaluatorResult(BaseModel):
    """Result from a single evaluator."""

    evaluator_type: str
    score: float  # 0-1
    passed: bool
    explanation: str = ""
    details: dict[str, str] = {}


class EvalResult(BaseModel):
    """Result for a single example."""

    id: str
    eval_run_id: str
    example_id: str
    input: dict[str, Any] = {}
    expected_output: dict[str, Any] = {}
    actual_output: dict[str, Any] = {}
    evaluator_results: dict[str, EvaluatorResult] = {}
    overall_score: float = 0.0
    passed: bool = False
    latency_ms: float = 0.0
    tokens_used: int = 0
    cost_usd: float = 0.0
    error: str = ""


class EvalSummary(BaseModel):
    """Summary statistics for an evaluation run."""

    overall_score: float = 0.0
    scores_by_evaluator: dict[str, float] = {}
    passed_count: int = 0
    failed_count: int = 0
    pass_rate: float = 0.0
    total_cost_usd: float = 0.0
    total_tokens: int = 0
    avg_latency_ms: float = 0.0


class EvalRun(BaseModel):
    """An evaluation run."""

    id: str
    name: str
    description: str = ""
    prompt_id: str = ""
    prompt_version: int = 0
    dataset_id: str = ""
    config: EvalConfig | None = None
    status: EvalRunStatus = EvalRunStatus.UNSPECIFIED
    error_message: str = ""
    total_examples: int = 0
    completed_examples: int = 0
    summary: EvalSummary | None = None
    created_at: datetime | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_by: str = ""
    metadata: dict[str, str] = {}

    @property
    def progress(self) -> float:
        """Get completion progress as a percentage."""
        if self.total_examples == 0:
            return 0.0
        return (self.completed_examples / self.total_examples) * 100


class EvaluatorParam(BaseModel):
    """Parameter definition for an evaluator."""

    name: str
    type: str
    description: str = ""
    required: bool = False
    default_value: str = ""


class Evaluator(BaseModel):
    """An available evaluator type."""

    type: str
    name: str
    description: str = ""
    params: list[EvaluatorParam] = []


class RunComparison(BaseModel):
    """Summary for comparing runs."""

    run_id: str
    prompt_version: str = ""
    overall_score: float = 0.0
    pass_rate: float = 0.0
    avg_latency_ms: float = 0.0
    total_cost_usd: float = 0.0


class ExampleComparison(BaseModel):
    """Comparison of a single example across runs."""

    example_id: str
    score_a: float = 0.0
    score_b: float = 0.0
    score_diff: float = 0.0
    regression: bool = False
